torch/_inductor/select_algorithm.py

```diff
@@ -1066,63 +1113,177 @@ def _jinja2_env():
         return None
 
 
+class GeneratedCodeCacheEntry(NamedTuple):
+    code: str
+    extra: str
+    events: list[Any]
+
+
+class GeneratedCodeCache:
+    def __init__(self, *args, **kwargs):
+        self._cache: dict[str, GeneratedCodeCacheEntry] = {}
+
+    def cache_clear(self) -> None:
+        self._cache.clear()
+
+    def __repr__(self):
+        return repr(self._cache)
+
+    def make_key(
+        self,
+        input_nodes: tuple[ir.StorageBox],
+        num_stages: int,
+        num_warps: int,
+        call_sizes: list[sympy.core.symbol.Symbol],
+        prefix_args: int,
+        suffix_args: int,
+        epilogue_fn: Optional[Callable[..., Any]],
+        subgraphs: Optional[list[ir.Buffer]],  # has to be none to cache
+        workspace_arg: Optional[WorkspaceArg],  # has to be none to cache
+        layout: ir.Layout,
+        num_consumer_groups: int,
+        num_buffers_warp_spec: int,
+        kwargs: dict[str, Any],
+    ) -> Optional[str]:
+        symbol_mapping = {}
+        next_symbol_index = 1
+
+        def normalize_symbols(x: sympy.core.Expr) -> str:
+            symbols = x.free_symbols
+            nonlocal next_symbol_index
+            original_string = str(x)
+            for s in symbols:
+                if str(s) not in symbol_mapping:
+                    symbol_mapping[str(s)] = f"_normalized_symbol{next_symbol_index}"
+                    next_symbol_index += 1
+                original_string = original_string.replace(
+                    str(s), symbol_mapping[str(s)]
+                )
+
+            return original_string
+
+        def normalize_list(ls: Sequence[sympy.core.Expr]) -> list[str]:
+            return [normalize_symbols(x) for x in ls]
+
+        def layout_key(layout: ir.Layout) -> str:
+            return repr(
+                [
+                    normalize_list(layout.size),
+                    normalize_list(layout.stride),
+                    layout.dtype,
+                    layout.device,
+                    normalize_symbols(layout.offset),
+                ]
+            )
+
+        if subgraphs is None and workspace_arg is None:
```

Comment:
would not need the workspace arg carve out if we just included full args

Reply:
I added the workspace arg in the state, is this needed for mm_template?